### PR TITLE
Merge POKER and POKER_VARIANT Events Into Addon

### PR DESCRIPTION
**Pull request by minin43**

### CODE REVIEW

File: lua/randomat2/cl_events/cl_poker.lua [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+        self.ExtraPokerHandFrame:Remove()
+        self.ExtraPokerHandFrame = nil
+    end
+
+    self.Players = {}
+    self.Hand = {}
+    self.ExtraHand = {}
+    self.IsPlaying = false
+    self.CurrentlyBetting = false
+    -- self.ShouldPlayStartSound = false
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker.lua]: Dead code

File: lua/randomat2/cl_events/cl_poker.lua [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    self.CloseButton:SetPos(0, 0)
+    self.CloseButton:SetSize(self.Close:GetWide(), self.Close:GetTall())
+    self.CloseButton:SetText("CLOSE")
+    self.CloseButton.CustomDoClick = function()
+        if self then
+            self:ClosePanel()
+        end
+    end
+
+    if self.ExtraPokerHandFrame and self.ExtraPokerHandFrame:IsValid() then
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker.lua:286]: If you wanted to reduce the math duplication (calling SetPos twice), you could do the math for the normal position and save the result, then modify it if the extra frame is present and call SetPos with the result once.

e.g.
```lua
local x = some math
local y = some math
if extra then
    y = y + some more math
end
SetPos(x, y)
```

File: lua/randomat2/cl_events/cl_poker.lua [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+    else
+        self.PokerMain:TemporaryMessage(ply:Nick() .."'s turn to bet!")
+    end
+
+    self.PokerPlayers:SetPlayerAsBetting(ply)
+    self.PokerHand:SetCanDiscard(false)
+end
+
+function EVENT:RegisterBet(ply, betType, betAmount, plySteamId)
+    if ply == LocalPlayer() then
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker.lua]: Can this `LocalPlayer()` be replaced with `self.Self`?

File: lua/randomat2/cl_events/cl_poker.lua [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    end
+end
+
+function EVENT:EndDiscard()
+    self.PokerMain:SetTimer(0)
+end
+
+function EVENT:RegisterWinner(winner, hand)
+    if winner then
+        if winner == LocalPlayer() then
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker.lua]: Can this `LocalPlayer()` be replaced with `self.Self`?

File: lua/randomat2/cl_events/cl_poker_vgui.lua [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+    if self.Player then
+        self:CalculateNameSize()
+    end
+end
+
+function PlayerCard:SetPlayer(ply)
+    if ply then
+        self.Player = ply:Nick()
+
+        if string.len(self.Player) > 12 then
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker_vgui.lua]: Might need to do this to support weird characters in player names
```suggestion
        if utf8.len(self.Player) > 12 then
```

File: lua/randomat2/cl_events/cl_poker_vgui.lua [hunk 6]
```diff
@@ -0,0 +1,942 @@
...
+
+    surface.SetTextPos(self:GetWide() * 0.5 - (self.BetWidth * 0.5), self:GetTall() * 0.5 - (self.BetHeight * 0.5) + 5)
+    surface.DrawText(self.BetText)
+
+    surface.SetFont("Trebuchet18")
+    surface.SetTextPos((self:GetWide() * 0.5) - (self.ActionWidth * 0.5) + 3, self:GetTall() - self.ActionHeight - 2)
+    surface.DrawText(self.ActionText)
+
+    if self.IsBetting then
+        -- surface.SetDrawColor(255, 255, 255)
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker_vgui.lua:169]: Dead code

File: lua/randomat2/cl_events/cl_poker_vgui.lua [hunk 7]
```diff
@@ -0,0 +1,942 @@
...
+            end
+            if ( funcFunction ) then pnl.DoClick = funcFunction end
+        
+            menuPnl:AddPanel( pnl )
+        
+            return pnl
+        end
+
+        for k, v in pairs( pnl.Choices ) do
+			local option = pnl.Menu:AddOption( v, function() pnl:ChooseOption( v, k ) end )
```
Comment (Malivil) [inline lua/randomat2/cl_events/cl_poker_vgui.lua]: Tabs -> spaces

File: lua/randomat2/events/sv_poker.lua [hunk 8]
```diff
@@ -0,0 +1,1136 @@
...
+local EVENT = {}
+local EVENT_VARIANT = {}
+local EVENT_REF = nil
+
+EVENT.Title = "A Round Of Yogscast Poker"
+EVENT.Description = "Only if the 9 of Diamonds touch!"
+EVENT.ExtDescription = "A round of 5-Card Draw Poker (no Texas Hold 'Em, for my sake), bet with your\nhealth. Up to 7 may play. Any pair, three, or four of a kind containing the 9 of\nDiamonds instantly wins."
+EVENT.id = "poker"
+EVENT.MinPlayers = 2
+EVENT.Type = EVENT_TYPE_DEFAULT
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua:35]: You don't need to specify this since it's the default

File: lua/randomat2/events/sv_poker.lua [hunk 9]
```diff
@@ -0,0 +1,1136 @@
...
+
+local EVENT = {}
+local EVENT_VARIANT = {}
+local EVENT_REF = nil
+
+EVENT.Title = "A Round Of Yogscast Poker"
+EVENT.Description = "Only if the 9 of Diamonds touch!"
+EVENT.ExtDescription = "A round of 5-Card Draw Poker (no Texas Hold 'Em, for my sake), bet with your\nhealth. Up to 7 may play. Any pair, three, or four of a kind containing the 9 of\nDiamonds instantly wins."
+EVENT.id = "poker"
+EVENT.MinPlayers = 2
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]:
```suggestion
EVENT.MinPlayers = {
    Min = 2
}
```

File: lua/randomat2/events/sv_poker.lua [hunk 10]
```diff
@@ -0,0 +1,1136 @@
...
+    local removedPlayers = {}
+    local playersToPlay = {}
+
+    -- If continuous play is enabled, return the previous player list, clean up any disconnected players, and add in as many new players as possible
+    if ConVars.EnableContinuousPlay:GetBool() and #self.ContinuousPlayers > 0 then
+        playersToPlay = table.Copy(self.ContinuousPlayers)
+
+        local playersToRemove = {}
+        for _, ply in ipairs(playersToPlay) do
+            if !IsValid(ply) or !ply:IsValid() or not ply:Alive() then
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: Consistency. Also "IsValid" already calls `ply:IsValid`
```suggestion
            if not IsValid(ply) or not ply:Alive() then
```

File: lua/randomat2/events/sv_poker.lua [hunk 11]
```diff
@@ -0,0 +1,1136 @@
...
+        end
+    end
+
+    return false
+end
+
+-- Called to register a player's bet (or lack thereof)
+function EVENT:RegisterPlayerBet(ply, bet, betAmount, forceBet)
+    if not self.Started then self:End() return end
+    if not ply:IsValid() then return end
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]:
```suggestion
    if not IsValid(ply) then return end
```

File: lua/randomat2/events/sv_poker.lua [hunk 12]
```diff
@@ -0,0 +1,1136 @@
...
+        PlayerFolds(ply)
+
+        if not EnoughPlayersRemaining() then
+            if CanDispenseWinnings() then
+                self:CalculateWinner()
+            else
+                net.Start("DeclareNoWinner")
+                net.Broadcast()
+
+                timer.Simple(GetDynamicRoundTimerValue("RoundStateMessage"), function()
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: Recommend against using simple timers because they are not stoppable if, for instance, the event is ended early.

File: lua/randomat2/events/sv_poker.lua [hunk 13]
```diff
@@ -0,0 +1,1136 @@
...
+                    self:End()
+                end)
+            end
+        end
+    end
+end
+
+function EVENT:EndBetting()
+    timer.Simple(GetDynamicRoundTimerValue("RoundStateMessage"), function()
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: Recommend against using simple timers because they are not stoppable if, for instance, the event is ended early.

File: lua/randomat2/events/sv_poker.lua [hunk 14]
```diff
@@ -0,0 +1,1136 @@
...
+function EVENT:RegisterPlayerDiscard(ply, discardsTable)
+    if not self.Started then self:End() return end
+
+    if not self.AcceptingDiscards then return end
+
+    for _, cardToRemove in ipairs(discardsTable) do
+        local toBeRemoved
+        for index, cardInHand in ipairs(ply.Cards) do
+            if cardToRemove.Rank == cardInHand.Rank and cardToRemove.Suit == cardInHand.Suit then
+                toBeRemoved = index
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua:508]:
```suggestion
                toBeRemoved = index
                break
```

File: lua/randomat2/events/sv_poker.lua [hunk 15]
```diff
@@ -0,0 +1,1136 @@
...
+            if bet > 0 then
+                if IsAllIn(bet) then
+                    ply:Kill()
+                else
+                    ply:SetHealth(math.max(1, ply:Health() - healthToLose))
+                    ply:SetMaxHealth(math.max(1, ply:GetMaxHealth() - healthToLose))
+                end
+            else
+                -- Seems we somehow entered this at some point? I think related to using Bots
+                ErrorNoHalt("Detected invalid bet", ply, bet)
```
Comment (minin43) [inline lua/randomat2/events/sv_poker.lua]: Forgot to remove this

File: lua/randomat2/events/sv_poker.lua [hunk 16]
```diff
@@ -0,0 +1,1136 @@
...
+end
+
+function EVENT:CompletePlayerDiscarding()
+    timer.Remove("AcceptDiscards")
+    self.AcceptingDiscards = false
+    self.HaveDiscarded = true
+
+    self:DealDeck(true)
+
+    timer.Simple(GetDynamicRoundTimerValue("RoundStateMessage"), function()
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: Recommend against using simple timers because they are not stoppable if, for instance, the event is ended early.

File: lua/randomat2/events/sv_poker.lua [hunk 17]
```diff
@@ -0,0 +1,1136 @@
...
+    else
+        net.Start("DeclareWinner")
+            net.WriteEntity(winner)
+            net.WriteString(hand)
+        net.Broadcast()
+
+        self:ApplyRewards(winner, hand)
+    end
+
+    timer.Simple(GetDynamicRoundTimerValue("RoundStateEnd"), function()
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: Recommend against using simple timers because they are not stoppable if, for instance, the event is ended early.

File: lua/randomat2/events/sv_poker.lua [hunk 18]
```diff
@@ -0,0 +1,1136 @@
...
+function EVENT:ContinuousPlay()
+    self:End(true)
+    
+    timer.Simple(0, function()
+        self:Begin()
+    end)
+end
+
+-- Called when an event is stopped. Used to do manual cleanup of processes started in the event.
+function EVENT:End(continuousEnd)
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua:836]: All timers should be removed

File: lua/randomat2/events/sv_poker.lua [hunk 19]
```diff
@@ -0,0 +1,1136 @@
...
+            })
+        end
+
+        EVENT_REF:RegisterPlayerDiscard(ply, cardsBeingDiscarded)
+    end
+end)
+
+--// Hooks
+
+function HandlePokerPlayerDeath(ply)
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]: This should be either local or namespaced to reduce global pollution

File: lua/randomat2/events/sv_poker.lua [hunk 20]
```diff
@@ -0,0 +1,1136 @@
...
+
+        prevPlayer.NextPlayer = nextPlayer
+        nextPlayer.PrevPlayer = prevPlayer
+
+        HandlePokerPlayerDeath(ply)
+    end
+end)
+
+hook.Add("PlayerSay", "LoganDebugCommands", function(ply, msg)
+    if EVENT_REF.Started and ply:IsSuperAdmin() then // ply:SteamID64() == "76561198029935530" then
```
Comment (Malivil) [inline lua/randomat2/events/sv_poker.lua]:
```suggestion
    if EVENT_REF.Started and ply:IsSuperAdmin() then
```

File: lua/randomat2/sh_events/sh_poker.lua [hunk 21]
```diff
@@ -0,0 +1,258 @@
+-- Logan Christianson
+
+ConVars = ConVars or {}
```
Comment (Malivil) [inline lua/randomat2/sh_events/sh_poker.lua]: I'm concerned about such a generic global. Any addon could use the same name and overwrite your stuff.

Consider putting all these globals (tables, properties, methods) in a namespace such as `POKER`

File: lua/randomat2/sh_events/sh_poker.lua [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+        return ""
+    elseif rank == Cards.ACE then
+        return "ace"
+    elseif rank == Cards.JACK then
+        return "jack"
+    elseif rank == Cards.QUEEN then
+        return "queen"
+    elseif rank == Cards.KING then
+        return "king"
+    else
```
Comment (Malivil) [inline lua/randomat2/sh_events/sh_poker.lua:121]: This is more of a stylistic thing (so you can ignore it) but you can skip the last `else` and just `return`.
e.g.
```lua
if a then
 return 1
elseif b then
 return 2
end
return 3
```

File: lua/randomat2/cl_events/cl_poker.lua [hunk 23]
```diff
@@ -0,0 +1,498 @@
...
+            self.PokerMain:PermanentMessage("You win! Getting your bonus health now!")
+
+            if ConVars.EnableYogsification:GetBool() and self.ShouldPlayStartSound then
```
Comment (minin43) [inline lua/randomat2/cl_events/cl_poker.lua]: Forgot to include ConVars.EnableRoundStateAudioCues:GetBool() check here and line 258

File: lua/randomat2/cl_events/cl_poker.lua (hunk 2 above)
Comment (minin43) [reply lua/randomat2/cl_events/cl_poker.lua:286]: In a Paint or Think function I'd be concerned, but since this is a one-off, I am not

File: lua/randomat2/cl_events/cl_poker_vgui.lua (hunk 5 above)
Comment (minin43) [reply lua/randomat2/cl_events/cl_poker_vgui.lua]: Didn't even consider this a possibility

File: lua/randomat2/cl_events/cl_poker_vgui.lua (hunk 6 above)
Comment (minin43) [reply lua/randomat2/cl_events/cl_poker_vgui.lua:169]: Icon that might get re-added at a later point

File: lua/randomat2/events/sv_poker.lua (hunk 8 above)
Comment (minin43) [reply lua/randomat2/events/sv_poker.lua:35]: But I like specifying

File: lua/randomat2/events/sv_poker.lua (hunk 14 above)
Comment (minin43) [reply lua/randomat2/events/sv_poker.lua:508]: I believe this would break functionality

Comment (Malivil) [reply lua/randomat2/events/sv_poker.lua:508]: You expect multiple entries in `ply.Cards` to match suit and rank?

File: lua/randomat2/sh_events/sh_poker.lua (hunk 21 above)
Comment (minin43) [reply lua/randomat2/sh_events/sh_poker.lua]: I'll rename the global so it's a bit less generic, but even a namespace name like that could still run into issues with someone else trying to write their own poker game

Comment (Malivil) [reply lua/randomat2/sh_events/sh_poker.lua]: Then call it something like `RandomatPoker` or `LoganPoker`. I'm more concerned about the global pollution with all the methods and properties.

File: lua/randomat2/sh_events/sh_poker.lua (hunk 22 above)
Comment (minin43) [reply lua/randomat2/sh_events/sh_poker.lua:121]: Totally just talking theory here: depending on the context, I may prefer the explicit if-else with each having separate returns when the point of the code block is an explicit split of functionality (if a, return 1, if b, return two, else, return 3). In other scenarios, I may skip defining it as an "else" block if that's the core functionality intended by the block, and the early returns are logically "early escapes", where we're accounting for some pre-existing condition altering the standard functionality in some way.

In other words, this would be a case statement if I wasn't writing Lua code.

File: lua/randomat2/events/sv_poker.lua (hunk 14 above)
Comment (minin43) [reply lua/randomat2/events/sv_poker.lua:508]: Oh, to break the inner loop.

I'm pretty sure I did that at one point, but I swear it was breaking out of BOTH loops and breaking functionality.
